src/docs/arch-fast-truncate.dox

```diff
@@ -0,0 +1,843 @@
+
+Meanwhile, the \c inst_updates field is kept independently until the transaction that
+created it is resolved.
+It in effect belongs to that transaction and is neither needed nor used by anything else,
+with one exception: eviction checks whether it is non-NULL before evicting the page.
+(Pages with an instantiated but uncommitted truncate cannot be evicted.)
+
+Because it is possible for the page to split between instantiation and transaction
+resolution, finding the updates created during instantiation to resolve them is
+problematic.
+The \c inst_updates field makes this possible.
+
+Upon resolution (either commit or rollback) the \c inst_updates field is discarded.
+
+@section ft_instantiation Instantiation
+
+Instantiation is the process by which an on-disk deleted page (ref in state \c
+WT_REF_DELETED) is converted into an in-memory page (ref in state \c WT_REF_MEM) with all
+its items explicitly deleted with tombstone updates.
+<i>Semantically, this is an identity transformation.</i>
+
+This occurs under two sets of circumstances: first, if a thread that cannot yet see the
+deletion tries to read from the deleted page; and second, if a search lands on the deleted
+page's portion of the namespace.
+It can happen either before or after the transaction that deleted the page is resolved.
+(But not at the same time; locking the ref prevents that.)
+
+(Note that for searches that are positioning to do updates, the instantiation is
+unavoidable; however, for searches that are only reading, it would be better to return \c
+WT_NOTFOUND without pointlessly instantiating the page.
+This is not currently implemented and as of this writing it is not entirely clear how
+involved or feasible such changes would be.
+Also note that this only applies to explicit searches and searches that are part of other
+cursor operations; the cursor next and previous operations do skip over deleted pages.)
+
+In either case the fact that the page is deleted is detected when we read the page.
+If the page has no address, or it has an address but the deletion is globally visible, we
+create a new in-memory page instead of reading the on-disk page.
+Otherwise, we read the on-disk page and call the instantiation code in \c bt_delete.c.
+
+The instantiation code iterates the page and adds a tombstone for every item.
+(To be precise, it adds a tombstone for every item that isn't already deleted; items that
+have a stop time do not need to be deleted again.)  The row-store version of this iterates
+the entries on the page and directly adds a tombstone to each update list.
+The column-store version uses a cursor and calls into \c col_modify.c to do it; this is
+because the data structures are considerably more complicated and updating them directly
+would require a lot of cut-and-paste code.
+
+The tombstones are tagged with WT_UPDATE_RESTORED_FAST_TRUNCATE.
+This is used by prepare to avoid trying to coalesce these updates with others to the same
```

Comment:
By `prepare`, do you mean `prepare_transaction`? If that's the case, can we be more explicit and add a `\c` as well?

Reply:
I meant "prepared transaction handling" but it's actually in `__wt_txn_prepare` and I might as well be explicit. Fixed in [f3a555c](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/f3a555cfd7ddc3655258d577632287a485c5676a)